Frameworks/ImageIO/CGImageDestination.mm

```diff
@@ -28,75 +885,358 @@ CGImageDestinationRef CGImageDestinationCreateWithDataConsumer(CGDataConsumerRef
+    RETURN_NULL_IF(!url);
+    
+    return (CGImageDestinationRef)[[ImageDestination alloc] initToDestination:count type:type data:NULL url:url];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Caveat
+ @Notes The current implementation supports common image file formats such as JPEG, TIFF, BMP, GIF, and PNG. 
+        Not all formats are supported. Also, the option for kCGImageDestinationBackgroundColor is currently not supported.
+        Not all image properties for all images are supported, and Apple writes many PNG properties that are not actually
+        part of the PNG specification to image metadata, which is not reproduced.
+        DPIWidth/DPIHeight are aliased to TIFF XDensity/YDensity. Both of these are only written if both X/Y values are specified.
+        For JPEG, the TIFF XDensity/YDensity values take precedence if both are present. For TIFF, it's DPIWidth/DPIHeight.
+        Also, if left blank, WIC will automatically set these values for TIFF to 96 DPI. WIC will not allow actually leaving blanks.
+        Worth noting, iOS seems to ignore the TIFF dictionary entirely if a JFIF dictionary is present. This is not replicated.
 */
 void CGImageDestinationAddImage(CGImageDestinationRef idst, CGImageRef image, CFDictionaryRef properties) {
-    UNIMPLEMENTED();
+    if (!idst || !image) {
+        return;
+    }
+
+    ImageDestination* imageDestination = (ImageDestination*)idst;
+    if (imageDestination.count >= imageDestination.maxCount) {
+        NSTraceInfo(TAG, @"Max number of images in destination exceeded\n");
+        return;
+    }
+
+    ComPtr<IWICBitmapFrameEncode> imageBitmapFrame;
+    IPropertyBag2* propertyBag = NULL;
+    
+    ComPtr<IWICBitmapEncoder> imageEncoder = imageDestination.idEncoder;
+
+    // If there is no Encoder, the destination has either been already finalized or not initialized yet, so return
+    if (!imageEncoder) {
+        NSTraceInfo(TAG, @"Destination object has no Encoder\n");
+        return;
+    }
+
+    HRESULT status = imageEncoder->CreateNewFrame(&imageBitmapFrame, &propertyBag);
+    if (!SUCCEEDED(status)) {
+        NSTraceInfo(TAG, @"CreateNewFrame failed with status=%x\n", status);
+        return;
+    }
+
+    // JPEG lossy compression property
+    // Note that while the scale goes from 0.0 to 1.0 on both Windows and iOS, WIC treats 0.0 as a much worse
+    // quality than iOS. In iOS, 0.0 looks relatively similar, while in Windows the image is heavily warped.
+    if (properties && CFDictionaryContainsKey(properties, kCGImageDestinationLossyCompressionQuality) &&
+            imageDestination.type == typeJPEG) {
+        PROPBAG2 option = { 0 };
+        std::wstring quality(L"ImageQuality");
+        option.pstrName = (wchar_t*)quality.c_str();
+        VARIANT varValue;    
+        VariantInit(&varValue);
+        varValue.vt = VT_R4;
+        varValue.bVal = [(id)CFDictionaryGetValue(properties, kCGImageDestinationLossyCompressionQuality) doubleValue];
```

Comment:
Yes, this was a bug I actually missed because I was only testing quality = 0.0, which an unitialized bVal would produce anyways. Thanks for catching it.